crates/matrix-sdk-ui/src/timeline/mod.rs

```diff
@@ -543,6 +546,40 @@ impl Timeline {
         SendAttachment::new(self, path.into(), mime_type, config)
     }
 
+    /// Redacts an event from the timeline.
+    ///
+    /// If it was a local event, this will *try* to cancel it, if it was not
+    /// being sent already. If the event was a remote event, then it will be
+    /// redacted by sending a redaction request to the server.
+    ///
+    /// Returns whether the redaction did happen. It can only return false for
+    /// local events that are being processed.
+    pub async fn redact(
+        &self,
+        event: &EventTimelineItem,
+        reason: Option<&str>,
+    ) -> Result<bool, RedactEventError> {
+        match &event.kind {
+            event_item::EventTimelineItemKind::Local(local) => {
+                if let Some(handle) = local.abort_handle.clone() {
+                    Ok(handle.abort().await)
+                } else {
+                    // No abort handle; theoretically unreachable for regular usage of the
+                    // timeline, but this may happen in testing contexts.
```

Comment:
For what it's worth, I tried introducing a Timeline-wide `enum AbortSendHandle` that had two variants, one for testing purposes that's a noop, and one that's the "real" `sending_queue::AbortSendHandle`, but then I noticed that for reactions, we have a custom method `toggle_reaction()` that should be used for removing (redacting) a reaction. So this would have required a third variant "NoThisIsAReactionRedaction", and error handling on top of this.

Rather, we should likely investigate how reactions work in the timeline, and fix those (there are observable UX errors related to reactions, and the state management could likely be simplified and reunited with other methods of the timeline).

Thanks for the suggestion, though!